### PR TITLE
Adds support for ASN1 Real BER encoding decoding

### PR DESCRIPTION
**Pull request by JanFellner**

### CODE REVIEW

File: test/asn1choice.spec.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+import { SchemaContext } from "../src";
+import { ETagClass } from "../src/TypeStore";
+
+/**
+ * Converts an array buffer to hex notation
+ *
+ * @param buffer - the buffer to convert
+ * @returns the buffer in hex string notation
+ */
+function buf2hex(buffer: ArrayBuffer): string {
```
Comment (microshine) [inline test/asn1choice.spec.ts:13]: I see you implemented some helping functions like `buf2hex`, `hex2buf` and use their copies in each test file. `pvtsutils` has `Convert` static class which implements this functionality, please use it

File: src/schema.ts [hunk 2]
```diff
@@ -16,397 +22,524 @@ export interface CompareSchemaSuccess {
...
+  public continueOnError: boolean;
+  // If the asn1 object is larger than the schema, this is not an error
+  public allowLargerThanSchema: boolean;
+}
+
+/**
+ * A helper object that is carried around while recursing through the schema
+ */
+export class SchemaContext {
+  // The path within the structure we are parsing (object:subobject:element)
```
Comment (microshine) [inline src/schema.ts:47]: use doc comment

File: tsconfig.json [hunk 3]
```diff
@@ -9,6 +9,7 @@
     "importHelpers": true,
     "noImplicitOverride": true,
     "skipLibCheck": true,
+    "sourceMap": true,
```
Comment (microshine) [inline tsconfig.json:12]: What do you use `sourceMap` for? We don't add source map files into the build folder and it's not needed for debuging

File: .gitignore [hunk 4]
```diff
@@ -1,8 +1,13 @@
...
-.vscode
 **/node_modules
 **/package-lock.json
 build
 .nyc_output
 coverage
-.DS_Store
\ No newline at end of file
+.DS_Store
+src/*.js
```
Comment (microshine) [inline .gitignore:8]: looks like you build JS files using `tsc` command. Please see this extension https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-mocha-test-adapter. It allows runnig  mocha tests without `sourceMap`, compiling and `launch.json` file

**Review by microshine — CHANGES_REQUESTED** [review]
@JanFellner thank you for your update. I'd like to include these features into the asn1js module.

I've requested some changes

This build works fine with pkijs and @peculiar/asn1-schema

**Comment by JanFellner** [comment]
Hi Stephan, 

Thanks for catching up.

I made so many changes in the repo and accidently setup the PR from the master branch.  So the core idea of adding asn real is blurred with so many other changes that it does not make sense to merge it right now.

To give you a little context. I am implementing asn.1 BER as communication layer using ROSE. Doing this I see several things that need to get adopted/improved in the lib. I've been working with the lib for the last 4 weeks and made several improvements. The changes you require are easy to adopt the question is if it makes sense to merge them all.

Things I have touches.
- adding real
- adding contextual optional params where the encoded data says a value is context specific and points to the schema which value (and therefor type) it is.
- fixes for repeated values as these have not been validated.
- after parsing the data against the schema the attribute names are taken over to the parsed data to have them accessible by name.
- the schema validation returns an array of errors giving the context where a parsing error has been seen. 
- and other tweaks to improve usability.

- all extensions i made are test covered. 
- all existing test are passing
- the addtional tests also explain the usage of the library. (how to handle optionals. repeated. Etc.) 

In case you think about taking those changes over I would place them as a new major version cause they are rather large in the end.

Would be great if you could check that on your side in a productive environment if it still does what it is expected to do.

The changes I made are matching to BER encoded objects by the snacc and objective systems compilers which we use today for our products.

If you think this does not really make sense for you, we would start to maintain an own fork. However you decide let me know 😉

Comments welcome.. 

Jan

**Comment by JanFellner** [comment]
I adopted all your recommondations but we decided to maintain our own version of the repo as the amount of changes especially in the schema validation concering implicit notated objects is rather large.
I do not want to put you into the need of maintaining stuff you don´t need.

**Pull request closed without merging**